core/domain/collection_services.py

```diff
@@ -0,0 +1,872 @@
+            collection_summary.collection_model_last_updated),
+        collection_model_created_on=(
+            collection_summary.collection_model_created_on)
+    )
+
+    collection_summary_model.put()
+
+
+def delete_collection_summary(collection_id, force_deletion=False):
+    """Delete an collection summary model."""
+
+    collection_models.CollectionSummaryModel.get(collection_id).delete()
+
+
+def save_new_collection_from_yaml(committer_id, yaml_content, collection_id):
+    collection = collection_domain.Collection.from_yaml(
+        collection_id, yaml_content)
+    commit_message = (
+        'New collection created from YAML file with title \'%s\'.'
+        % collection.title)
+
+    _create_collection(committer_id, collection, commit_message, [{
+        'cmd': CMD_CREATE_NEW,
+        'title': collection.title,
+        'category': collection.category,
+    }])
+
+    return collection
+
+
+def delete_demo(collection_id):
+    """Deletes a single demo collection."""
+    if not (0 <= int(collection_id) < len(feconf.DEMO_COLLECTIONS)):
+        raise Exception('Invalid demo collection id %s' % collection_id)
+
+    collection = get_collection_by_id(collection_id, strict=False)
+    if not collection:
+        logging.info('Collection with id %s was not deleted, because it '
+                     'does not exist.' % collection_id)
+    else:
+        delete_collection(
+            feconf.SYSTEM_COMMITTER_ID, collection_id, force_deletion=True)
+
+
+def load_demo(collection_id):
+    """Loads a demo collection.
+
+    The resulting collection will have version 2 (one for its initial
+    creation and one for its subsequent modification.)
+    """
+    delete_demo(collection_id)
+
+    if not collection_domain.Collection.is_demo_collection_id(collection_id):
+        raise Exception('Invalid demo collection id %s' % collection_id)
+
+    demo_filepath = os.path.join(
+        feconf.SAMPLE_COLLECTIONS_DIR,
+        feconf.DEMO_COLLECTIONS[int(collection_id)])
+
+    if demo_filepath.endswith('yaml'):
+        yaml_content = utils.get_file_contents(demo_filepath)
+    else:
+        raise Exception('Unrecognized file path: %s' % demo_filepath)
+
+    collection = save_new_collection_from_yaml(
+        feconf.SYSTEM_COMMITTER_ID, yaml_content, collection_id)
+
+    rights_manager.publish_collection(
+        feconf.SYSTEM_COMMITTER_ID, collection_id)
+
+    index_collections_given_ids([collection_id])
+
+    # Now, load all of the demo explorations that are part of the collection.
+    for collection_node in collection.nodes:
+        exp_id = collection_node.exploration_id
+        # Only load the demo exploration if it is not yet loaded.
+        if exp_services.get_exploration_by_id(exp_id, strict=False) is None:
+            exp_services.load_demo(exp_id)
+
+    logging.info('Collection with id %s was loaded.' % collection_id)
+
+
+# TODO(bhenning): Cleanup search logic and abstract it between explorations and

```

Comment:
Definitely +1 to unifying the search functionality for explorations and collections into a separate search_services (not activity_services) file. I don't think I have strong feelings with regards to having one or two indexes. (We could even postpone the decision and have three -- 'explorations', 'collections', 'all'.)

> If I understand this suggestion correctly, we would instead have a search_manager which has an actor class for the index and all the search functionality which acts upon the index. Then, the search functions in both collection_services and exp_services would just use that manager with their respective indexes. The call sites for exploration search functions wouldn't change. The unified ActvitySummary would still live in activity_domain.py.

Not quite, I think -- the index wouldn't literally have or be an 'actor', it's just the design pattern of the Actor class that I was referring to. So there would be a class like this one:

    class SearchIndex(object):
        """Domain object for a search index."""
        def __init__(self, name):
            self.name = name

        def ....